pkg/vgmanager/vgmanager_controller.go

```diff
@@ -76,6 +83,47 @@ func (r *VGReconciler) reconcile(ctx context.Context, req ctrl.Request) (ctrl.Re
 		return ctrl.Result{}, err
 	}
 
+	if vgFinalizer == "" {
+		node := &corev1.Node{}
+		err = r.Client.Get(ctx, types.NamespacedName{Name: nodeName}, node)
+		if err != nil {
+			r.Log.Error(err, "failed to get node object", "node name", nodeName)
+			return ctrl.Result{}, err
+		}
+		vgFinalizer = fmt.Sprintf("node-uid/%s", node.UID)
+	}
+
+	// cleanup on current node when CR is deleted
+	if !lvmCluster.DeletionTimestamp.IsZero() {
+		if controllerutil.ContainsFinalizer(lvmCluster, vgFinalizer) {
+			// TODO: any possibility to make vgCleanup async?
+			err := vgCleanupAll(r, ctx)
+			if err != nil {
+				r.Log.Error(err, "failed to perform VG Cleanup", "node", nodeName)
+				return ctrl.Result{}, err
+			}
+
+			// remove finalizer after cleanup
+			controllerutil.RemoveFinalizer(lvmCluster, vgFinalizer)
+			if err := r.Client.Update(ctx, lvmCluster); err != nil {
+				r.Log.Error(err, "failed to remove vg finalizer", "node", nodeName, "LVMCluster", lvmCluster.Name, "finalizer", vgFinalizer)
+				return ctrl.Result{}, err
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// set node uid as finalizer on CR
+	if !controllerutil.ContainsFinalizer(lvmCluster, vgFinalizer) {
+		controllerutil.AddFinalizer(lvmCluster, vgFinalizer)
```

Comment:
If each vgManager daemonset pod is adding its own finalizer to the LVMCluster CR, then we might have issues in cleaning up of the finalizers. For example: if NodeSelector of the Vgmanager daemonset is updated, we might have stale finalizers in the LVMCluster CR finalizer.